networking/eth2/src/testFixtures/java/tech/pegasys/teku/networking/eth2/peers/RespondingEth2Peer.java

```diff
@@ -0,0 +1,441 @@
+            () ->
+                blockRoots.stream()
+                    .map(this::findBlockByRoot)
+                    .flatMap(Optional::stream)
+                    .collect(Collectors.toList()));
+
+    return createPendingRequest(handler);
+  }
+
+  @Override
+  public SafeFuture<Optional<SignedBeaconBlock>> requestBlockBySlot(final UInt64 slot) {
+    final PendingBlockRequestHandler<Optional<SignedBeaconBlock>> handler =
+        PendingBlockRequestHandler.createForSingleBlockRequest(
+            () -> Optional.ofNullable(chain.getBlockAtSlot(slot)));
+
+    return createPendingRequest(handler);
+  }
+
+  @Override
+  public SafeFuture<Optional<SignedBeaconBlock>> requestBlockByRoot(final Bytes32 blockRoot) {
+    final PendingBlockRequestHandler<Optional<SignedBeaconBlock>> handler =
+        PendingBlockRequestHandler.createForSingleBlockRequest(() -> findBlockByRoot(blockRoot));
+
+    return createPendingRequest(handler);
+  }
+
+  private <T> SafeFuture<T> createPendingRequest(PendingBlockRequestHandler<T> handler) {
+    final PendingBlockRequestHandler<T> filteredHandler =
+        PendingBlockRequestHandler.filterRequest(handler, blockRequestFilter);
+    final PendingBlockRequest<T> request = new PendingBlockRequest<>(filteredHandler);
+
+    pendingRequests.add(request);
+    return request.getFuture();
+  }
+
+  @Override
+  public SafeFuture<MetadataMessage> requestMetadata() {
+    return SafeFuture.completedFuture(MetadataMessage.createDefault());
+  }
+
+  @Override
+  public <I extends RpcRequest, O> SafeFuture<O> requestSingleItem(
+      final Eth2RpcMethod<I, O> method, final I request) {
+    return SafeFuture.failedFuture(new UnsupportedOperationException());
+  }
+
+  @Override
+  public boolean wantToReceiveObjects(
+      final ResponseCallback<SignedBeaconBlock> callback, final long objectCount) {
+    return true;
+  }
+
+  @Override
+  public boolean wantToMakeRequest() {
+    return true;
+  }
+
+  @Override
+  public SafeFuture<UInt64> sendPing() {
+    return SafeFuture.completedFuture(UInt64.ONE);
+  }
+
+  @Override
+  public int getOutstandingPings() {
+    return 0;
+  }
+
+  @Override
+  public NodeId getId() {
+    return nodeId;
+  }
+
+  @Override
+  public PeerAddress getAddress() {
+    return new PeerAddress(nodeId);
+  }
+
+  @Override
+  public boolean isConnected() {
+    return !disconnected;
+  }
+
+  @Override
+  public void disconnectImmediately(
+      final Optional<DisconnectReason> reason, final boolean locallyInitiated) {
+    disconnect(reason, locallyInitiated);
+  }
+
+  @Override
+  public SafeFuture<Void> disconnectCleanly(final DisconnectReason reason) {
+    disconnect(Optional.of(reason), true);
+    return SafeFuture.COMPLETE;
+  }
+
+  private void disconnect(Optional<DisconnectReason> reason, boolean locallyInitiated) {
+    disconnected = true;
+    disconnectSubscribers.deliver(
+        PeerDisconnectedSubscriber::onDisconnected, reason, locallyInitiated);
```

Comment:
May be this way to avoid adding a new method to `Subscribers`? 
```suggestion
    disconnectSubscribers
        .forEach(subscriber -> subscriber.onDisconnected(reason, locallyInitiated));
```